File: scripts/plot_damage_profile.py
#!/usr/bin/env python3
"""
plot_damage_profile.py

Plot absolute-position aDNA damage profiles from adna_damage_estimate.py output.

X axis: position from read end (0 = terminal k-mer)
Y axis: fraction of unclassified k-mers (%)

One panel pair (5' end + 3' end) per taxid. Multiple samples can be overlaid
on the same axes by passing --profile and --global multiple times.

Usage:
  # Single sample
  python plot_damage_profile.py \\
    --profile sample_profile.tsv \\
    --global  sample_global.tsv \\
    --stats   sample_stats.tsv \\
    --output  damage_absolute.pdf \\
    [--taxids 1649845 632] \\
    [--max-pos 20]

  # Two samples overlaid (e.g. damaged vs undamaged)
  python plot_damage_profile.py \\
    --profile dam1_profile.tsv dam0_profile.tsv \\
    --global  dam1_global.tsv  dam0_global.tsv \\
    --stats   dam1_stats.tsv   dam0_stats.tsv \\
    --label   "aDNA damage"    "No damage" \\
    --output  damage_comparison.pdf \\
    [--taxids 1649845]

  --stats is optional but recommended: when provided, the adaptive plateau window
  and per-end p-values from adna_damage_estimate.py are used directly.
"""

import argparse
import sys
from pathlib import Path

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.backends.backend_pdf import PdfPages
import numpy as np
import pandas as pd


DEFAULT_COLORS = ["#d6604d", "#2166ac", "#4dac26", "#8073ac"]
DEFAULT_STYLES = ["-", "--", "-.", ":"]


def load_sample(profile_path, global_path, stats_path=None):
    profile = pd.read_csv(profile_path, sep="\t")
    gdf     = pd.read_csv(global_path,  sep="\t")
    sdf     = pd.read_csv(stats_path,   sep="\t") if stats_path else None
    return profile, gdf, sdf


def resolve_keys(gdfs, requested_taxids, requested_species, pvalue_threshold=0.05):
    """
    Return an ordered list of keys to plot — either int taxids or str species names.

    Explicit --taxids / --species take priority.  Otherwise auto-select all keys
    with significant evidence for aDNA damage (damage_score > 0 and
    damage_pvalue < pvalue_threshold in at least one sample), sorted by mean
    damage score descending.

    Falls back to top-6 by score when the global file predates p-value output.
    """
    if requested_species:
        return requested_species
    if requested_taxids:
        return requested_taxids

    combined = pd.concat(gdfs, ignore_index=True)
    taxid_present = (
        "taxid" in combined.columns
        and combined["taxid"].notna().any()
    )
    group_col = "taxid" if taxid_present else "species_name"

    if "damage_pvalue" in combined.columns:
        # keep keys that have at least one significant, positive-score row
        sig_mask = (
            combined["damage_score"].gt(0) &
            combined["damage_pvalue"].lt(pvalue_threshold)
        )
        sig_keys = set(combined.loc[sig_mask, group_col].dropna())
        candidates = combined[combined[group_col].isin(sig_keys)]
        keys = (
            candidates.groupby(group_col)["damage_pvalue"]
            .mean().dropna()
            .sort_values(ascending=True)
            .index.tolist()
        )
        if not keys:
            print(
                f"No taxa pass p < {pvalue_threshold} with positive damage score. "
                "Falling back to top 6 by score.",
                file=sys.stderr,
            )
        else:
            print(
                f"Auto-selected {len(keys)} taxa with damage_pvalue < {pvalue_threshold}.",
                file=sys.stderr,
            )
            return keys

    # fallback: no p-value column or nothing passed filter
    return (
        combined.groupby(group_col)["damage_score"]
        .mean().dropna()
        .sort_values(ascending=False)
        .head(3).index.tolist()
    )


def plot_end(ax, profiles, gdfs, stats_dfs, labels, colors, styles,
             key, end, max_pos, plateau_start, plateau_end):
    key_col = "species_name" if isinstance(key, str) else "taxid"

    for samp_idx, (profile, gdf, sdf, label, color, style) in enumerate(
        zip(profiles, gdfs, stats_dfs if stats_dfs else [None] * len(profiles),
            labels, colors, styles)
    ):
        sub = (profile[(profile[key_col] == key) & (profile["end"] == end)]
               .query("position < @max_pos")
               .sort_values("position"))
        if sub.empty:
            continue

        # per-sample plateau window: from stats if available, else CLI fallback
        ps, pe = plateau_start, plateau_end
        if sdf is not None:
            row = sdf[(sdf[key_col] == key) & (sdf["end"] == end)]
            if not row.empty and "plateau_pos_start" in sdf.columns:
                ps = int(row["plateau_pos_start"].iloc[0])
                pe = int(row["plateau_pos_end"].iloc[0])

        # score and p-value: prefer per-end stats, fall back to global
        pvalue = np.nan
        if sdf is not None:
            s = sdf[(sdf[key_col] == key) & (sdf["end"] == end)]
            score   = s["damage_score"].values[0]  if len(s) else np.nan
            n_reads = int(s["n_reads"].values[0])  if len(s) else 0
            if "damage_pvalue" in s.columns and len(s):
                pvalue = s["damage_pvalue"].values[0]
        else:
            g = gdf[gdf[key_col] == key]
            score_col = "damage_score" if end == "5prime" else "damage_score_3prime"
            score   = g[score_col].values[0]      if len(g) else np.nan
            n_reads = int(g["n_reads"].values[0]) if len(g) else 0

        lbl = f"{label}  (n={n_reads:,}, Δ={score*100:.2f}%"
        if not np.isnan(pvalue):
            lbl += f", p={pvalue:.1e}"
        lbl += ")"
        ax.plot(sub["position"], sub["frac_unclassified"] * 100,
                color=color, ls=style, lw=2, marker="o", markersize=3.5, label=lbl)

        # per-sample plateau: coloured vertical span + matching horizontal line
        ax.axvspan(ps, pe, alpha=0.08, color=color, lw=0)
        plateau = (sub[sub["position"].between(ps, pe)]
                   ["frac_unclassified"].mean() * 100)
        ax.axhline(plateau, color=color, lw=0.9, ls=":", alpha=0.6,
                   label=f"Plateau {ps}–{pe} ({plateau:.2f}%)")
    ax.set_title("5′ end" if end == "5prime" else "3′ end", fontsize=9)
    ax.set_xlabel("Position from read end", fontsize=8)
    ax.set_ylabel("Unclassified k-mers (%)", fontsize=8)
    ax.set_xlim(-0.5, max_pos - 0.5)
    ax.set_ylim(bottom=0)
    ax.legend(fontsize=7.5, frameon=False)
    ax.spines[["top", "right"]].set_visible(False)
    ax.tick_params(labelsize=7)


def make_plot(profiles, gdfs, stats_dfs, labels, colors, styles,
              keys, max_pos, plateau_start, plateau_end, output_path):
    footer = (
        "Rising tail beyond plateau reflects read-length artifact "
        "(short reads: distant positions approach the opposite end)."
    )
    with PdfPages(output_path) as pdf:
        for key in keys:
            if isinstance(key, str):
                page_title = key
            else:
                name = ""
                for gdf in gdfs:
                    m = gdf[gdf["taxid"] == key]
                    if not m.empty and "species_name" in gdf.columns:
                        name = str(m["species_name"].values[0]); break
                page_title = f"taxid {key}" + (f" — {name}" if name else "")

            fig, axes = plt.subplots(1, 2, figsize=(11, 4), squeeze=False)
            fig.suptitle(
                f"aDNA damage profiles — {page_title}",
                fontsize=11,
            )
            for col, end in enumerate(["5prime", "3prime"]):
                plot_end(axes[0][col], profiles, gdfs, stats_dfs, labels, colors, styles,
                         key, end, max_pos, plateau_start, plateau_end)
            fig.text(0.5, -0.02, footer,
                     ha="center", fontsize=8, color="gray", style="italic")
            fig.tight_layout()
            pdf.savefig(fig, bbox_inches="tight", dpi=150)
            plt.close(fig)

    print(f"Saved {output_path} ({len(keys)} page(s))", file=sys.stderr)


def parse_args():
    p = argparse.ArgumentParser(description=__doc__,
                                formatter_class=argparse.RawDescriptionHelpFormatter)
    p.add_argument("--profile", nargs="+", required=True,
                   help="*_profile.tsv file(s) from adna_damage_estimate.py")
    p.add_argument("--global",  nargs="+", required=True, dest="global_",
                   help="*_global.tsv file(s) from adna_damage_estimate.py")
    p.add_argument("--stats",   nargs="+", default=[],
                   help="*_stats.tsv file(s) — enables adaptive plateau shading and p-values")
    p.add_argument("--label",   nargs="+", default=[])
    p.add_argument("--output",  required=True)
    p.add_argument("--taxids",  type=int, nargs="+", default=[],
                   help="Integer taxids to plot")
    p.add_argument("--species", nargs="+", default=[],
                   help="Species names to plot (for species-aggregated profiles)")
    p.add_argument("--hits",       default=None,
                   help="all_samples.hits.tsv — use hit species for this sample as key list")
    p.add_argument("--sample-id",  default=None,
                   help="Sample ID to filter from --hits table")
    p.add_argument("--max-pos",          type=int, default=20)
    p.add_argument("--plateau-start",    type=int, default=3,
                   help="Fallback plateau start when --stats not provided (default: 3)")
    p.add_argument("--plateau-end",      type=int, default=10,
                   help="Fallback plateau end when --stats not provided (default: 10)")
    p.add_argument("--pvalue-threshold", type=float, default=0.05,
                   help="Max damage_pvalue for auto-selection of taxa (default: 0.05)")
    return p.parse_args()


def main():
    args = parse_args()
    if len(args.profile) != len(args.global_):
        print("ERROR: --profile and --global must match in count", file=sys.stderr)
        sys.exit(1)
    if args.stats and len(args.stats) != len(args.profile):
        print("ERROR: --stats count must match --profile count", file=sys.stderr)
        sys.exit(1)
    stats_paths = args.stats if args.stats else [None] * len(args.profile)
    loaded = [load_sample(p, g, s)
              for p, g, s in zip(args.profile, args.global_, stats_paths)]
    profiles  = [x[0] for x in loaded]
    gdfs      = [x[1] for x in loaded]
    stats_dfs = [x[2] for x in loaded]
    n = len(profiles)
    labels = args.label if len(args.label) == n else [f"Sample {i+1}" for i in range(n)]
    colors = [DEFAULT_COLORS[i % len(DEFAULT_COLORS)] for i in range(n)]
    styles = [DEFAULT_STYLES[i % len(DEFAULT_STYLES)] for i in range(n)]
    # --hits overrides auto-selection: use species from the hit table for this sample
    hit_species: list[str] = []
    if args.hits and args.sample_id:
        try:
            hdf = pd.read_csv(args.hits, sep="\t", dtype={"sample_id": str})
            hit_species = (
                hdf[hdf["sample_id"] == str(args.sample_id)]["species_name"]
                .dropna().unique().tolist()
            )
        except (FileNotFoundError, pd.errors.EmptyDataError):
            pass

    if hit_species:
        keys = hit_species
    else:
        keys = resolve_keys(gdfs, args.taxids, args.species, args.pvalue_threshold)

    if not keys:
        print("No taxa found. Use --hits, --taxids, or --species to specify targets.",
              file=sys.stderr)
        from matplotlib.backends.backend_pdf import PdfPages
        with PdfPages(args.output) as _pdf:
            pass
        sys.exit(0)
    make_plot(profiles, gdfs, stats_dfs, labels, colors, styles,
              keys, args.max_pos, args.plateau_start, args.plateau_end, args.output)


if __name__ == "__main__":
    main()

File: scripts/plot_damage_profile.R
#!/usr/bin/env Rscript
# ---------------------------------------------------------------------------
# plot_damage_profile.R
#
# Absolute-position aDNA damage profiles from workflow damage TSV outputs.
#
#   X axis : position from read end (0 = terminal k-mer)
#   Y axis : fraction of unclassified k-mers (%)
#
# One page per taxon, 5' and 3' stacked in one column on a shared y range.
# Line colour follows the mapDamage convention: 5' red, 3' blue. Multiple samples can be
# overlaid by passing --profile/--global/--stats more than once.
#
# --stats is optional but recommended: it supplies the adaptive plateau window
# and per-end p-values computed by aggregate_sample.py.
# ---------------------------------------------------------------------------

suppressPackageStartupMessages({
  library(ggplot2)
  library(readr)
  library(dplyr)
})

script_dir <- function() {
  a <- commandArgs(trailingOnly = FALSE)
  f <- sub("^--file=", "", a[grepl("^--file=", a)])
  if (length(f)) dirname(normalizePath(f)) else "."
}
source(file.path(script_dir(), "plot_common.R"))

SAMPLE_COLOURS <- c("#d6604d", "#2166ac", "#4dac26", "#8073ac")
SAMPLE_LINETYPES <- c("solid", "dashed", "dotdash", "dotted")

SPEC <- list(
  profile             = list(type = "character", nargs = "+", default = character(0)),
  global              = list(type = "character", nargs = "+", default = character(0)),
  stats               = list(type = "character", nargs = "+", default = character(0)),
  label               = list(type = "character", nargs = "+", default = character(0)),
  output              = list(type = "character", nargs = 1,   default = NULL),
  taxids              = list(type = "integer",   nargs = "+", default = integer(0)),
  species             = list(type = "character", nargs = "+", default = character(0)),
  hits                = list(type = "character", nargs = 1,   default = NULL),
  sample_id           = list(type = "character", nargs = 1,   default = NULL),
  hits_required_flags = list(type = "character", nargs = "+", default = character(0)),
  max_keys            = list(type = "integer",   nargs = 1,   default = 200L),
  max_pos             = list(type = "integer",   nargs = 1,   default = 20L),
  plateau_start       = list(type = "integer",   nargs = 1,   default = 3L),
  plateau_end         = list(type = "integer",   nargs = 1,   default = 10L),
  pvalue_threshold    = list(type = "double",    nargs = 1,   default = 0.05)
)

args <- parse_cli(commandArgs(trailingOnly = TRUE), SPEC)
stopifnot(length(args$profile) > 0, !is.null(args$output))

if (length(args$profile) != length(args$global)) {
  stop("--profile and --global must match in count", call. = FALSE)
}
if (length(args$stats) > 0 && length(args$stats) != length(args$profile)) {
  stop("--stats count must match --profile count", call. = FALSE)
}

n_samp <- length(args$profile)
labels <- if (length(args$label) == n_samp) args$label else paste("Sample", seq_len(n_samp))
colours <- SAMPLE_COLOURS[(seq_len(n_samp) - 1) %% length(SAMPLE_COLOURS) + 1]
ltypes  <- SAMPLE_LINETYPES[(seq_len(n_samp) - 1) %% length(SAMPLE_LINETYPES) + 1]

read_any <- function(p) suppressWarnings(readr::read_tsv(
  p, col_types = readr::cols(.default = readr::col_guess(),
                             species_name = readr::col_character()),
  progress = FALSE
))

profiles <- lapply(args$profile, read_any)
gdfs     <- lapply(args$global, read_any)
sdfs     <- if (length(args$stats) > 0) lapply(args$stats, read_any) else vector("list", n_samp)

# ---------------------------------------------------------------------------
# Which taxa to plot
# ---------------------------------------------------------------------------

resolve_keys <- function(gdfs, taxids, species, pthr) {
  if (length(species) > 0) return(species)
  if (length(taxids) > 0) return(taxids)

  combined <- dplyr::bind_rows(gdfs)
  has_taxid <- "taxid" %in% names(combined) && any(!is.na(combined$taxid))
  key_col <- if (has_taxid) "taxid" else "species_name"

  if ("damage_pvalue" %in% names(combined)) {
    sig <- combined$damage_score > 0 & combined$damage_pvalue < pthr
    sig_keys <- unique(stats::na.omit(combined[[key_col]][sig]))
    if (length(sig_keys) > 0) {
      cand <- combined[combined[[key_col]] %in% sig_keys, , drop = FALSE]
      agg <- aggregate(cand$damage_pvalue, by = list(key = cand[[key_col]]),
                       FUN = function(v) mean(v, na.rm = TRUE))
      agg <- agg[order(agg$x), , drop = FALSE]
      message("Auto-selected ", nrow(agg), " taxa with damage_pvalue < ", pthr, ".")
      return(agg$key)
    }
    message("No taxa pass p < ", pthr, " with positive damage score. ",
            "Falling back to top 3 by score.")
  }
  agg <- aggregate(combined$damage_score, by = list(key = combined[[key_col]]),
                   FUN = function(v) mean(v, na.rm = TRUE))
  agg <- agg[order(-agg$x), , drop = FALSE]
  utils::head(agg$key, 3)
}

use_hits <- !is.null(args$hits) && !is.null(args$sample_id)
if (use_hits) {
  sel <- select_hit_species(args$hits, args$sample_id,
                            resolve_required_hit_flags(args$hits_required_flags),
                            args$max_keys)
  keys <- sel$species
  if (length(keys) == 0L) {
    message("WARNING: no hit species matched all required hit flags for this sample.")
  }
  if (sel$n_truncated > 0L) {
    message("WARNING: selected hit species truncated to ", length(keys),
            " (dropped ", sel$n_truncated, " by --max-keys).")
  }
} else {
  keys <- resolve_keys(gdfs, args$taxids, args$species, args$pvalue_threshold)
  if (length(args$taxids) == 0 && length(args$species) == 0 &&
      args$max_keys > 0 && length(keys) > args$max_keys) {
    message("WARNING: auto-selected taxa truncated to ", args$max_keys,
            " (dropped ", length(keys) - args$max_keys, " by --max-keys).")
    keys <- keys[seq_len(args$max_keys)]
  }
}

# drop keys with no profile rows anywhere
if (length(keys) > 0 && is.character(keys)) {
  available <- unique(unlist(lapply(profiles, function(p)
    if ("species_name" %in% names(p)) stats::na.omit(as.character(p$species_name)) else character(0))))
  if (length(available) > 0) {
    kept <- keys[keys %in% available]
    if (length(kept) < length(keys)) {
      message("WARNING: dropped ", length(keys) - length(kept),
              " selected species with no profile rows.")
    }
    keys <- kept
  }
}

if (length(keys) == 0L) {
  write_empty_plot(
    args$output,
    "No damage-profile plots were generated",
    "No taxa were selected from --hits or auto-selection criteria."
  )
  quit(save = "no", status = 0)
}

# ---------------------------------------------------------------------------
# Per-page plot
# ---------------------------------------------------------------------------

END_LEVELS <- c("5prime", "3prime")
END_LABELS <- c(`5prime` = "5′ end", `3prime` = "3′ end")

build_page <- function(key) {
  key_col <- if (is.character(key)) "species_name" else "taxid"

  lines <- list(); spans <- list(); plats <- list()

  for (i in seq_len(n_samp)) {
    prof <- profiles[[i]]; sdf <- sdfs[[i]]; gdf <- gdfs[[i]]
    for (e in END_LEVELS) {
      sub <- prof[prof[[key_col]] == key & prof$end == e & prof$position < args$max_pos, , drop = FALSE]
      if (nrow(sub) == 0L) next
      sub <- sub[order(sub$position), , drop = FALSE]

      ps <- args$plateau_start; pe <- args$plateau_end
      score <- NA_real_; n_reads <- 0L; pval <- NA_real_
      if (!is.null(sdf)) {
        row <- sdf[sdf[[key_col]] == key & sdf$end == e, , drop = FALSE]
        if (nrow(row) > 0L) {
          if (all(c("plateau_pos_start", "plateau_pos_end") %in% names(row))) {
            ps <- as.integer(row$plateau_pos_start[1]); pe <- as.integer(row$plateau_pos_end[1])
          }
          score <- row$damage_score[1]
          n_reads <- as.integer(row$n_reads[1])
          if ("damage_pvalue" %in% names(row)) pval <- row$damage_pvalue[1]
        }
      } else {
        g <- gdf[gdf[[key_col]] == key, , drop = FALSE]
        sc <- if (e == "5prime") "damage_score" else "damage_score_3prime"
        if (nrow(g) > 0L) { score <- g[[sc]][1]; n_reads <- as.integer(g$n_reads[1]) }
      }

      lbl <- sprintf("%s  (n=%s, Δ=%.2f%%", labels[i],
                     format(n_reads, big.mark = ","), score * 100)
      if (!is.na(pval)) lbl <- paste0(lbl, sprintf(", p=%.1e", pval))
      lbl <- paste0(lbl, ")")

      lines[[length(lines) + 1L]] <- data.frame(
        end = e, position = sub$position,
        pct = sub$frac_unclassified * 100,
        series = lbl, sample_i = i, stringsAsFactors = FALSE
      )
      spans[[length(spans) + 1L]] <- data.frame(
        end = e, xmin = ps, xmax = pe, sample_i = i, stringsAsFactors = FALSE
      )
      inwin <- sub$position >= ps & sub$position <= pe
      plats[[length(plats) + 1L]] <- data.frame(
        end = e,
        yint = mean(sub$frac_unclassified[inwin], na.rm = TRUE) * 100,
        series = lbl, sample_i = i, stringsAsFactors = FALSE
      )
    }
  }

  if (length(lines) == 0L) return(NULL)
  lines <- dplyr::bind_rows(lines)
  spans <- dplyr::bind_rows(spans)
  plats <- dplyr::bind_rows(plats)
  lines$end <- factor(lines$end, levels = END_LEVELS)
  spans$end <- factor(spans$end, levels = END_LEVELS)
  plats$end <- factor(plats$end, levels = END_LEVELS)

  # Per-end read counts and damage statistics are annotated inside each panel
  # rather than carried in the series label: the label is built per
  # (sample, end), so with a single sample it produced one legend entry per end
  # for what is one series. A facet strip is too narrow for them when rotated.
  ann <- do.call(rbind, lapply(END_LEVELS, function(e) {
    lab <- unique(lines$series[lines$end == e])
    if (length(lab) == 0L) return(NULL)
    data.frame(end = e, label = paste(lab, collapse = "\n"), stringsAsFactors = FALSE)
  }))
  if (!is.null(ann)) ann$end <- factor(ann$end, levels = END_LEVELS)

  # mapDamage convention: 5' in red, 3' in blue. Colour therefore encodes the
  # read end; sample identity is carried by linetype when several are overlaid.
  end_pal <- c(`5prime` = "#d6604d", `3prime` = "#2166ac")

  sample_levels <- labels[sort(unique(lines$sample_i))]
  lty <- ltypes[sort(unique(lines$sample_i))]
  names(lty) <- sample_levels
  lines$sample <- factor(labels[lines$sample_i], levels = sample_levels)
  plats$sample <- factor(labels[plats$sample_i], levels = sample_levels)
  spans$sample <- factor(labels[spans$sample_i], levels = sample_levels)

  page_title <- if (is.character(key)) key else {
    nm <- ""
    for (g in gdfs) {
      m <- g[g$taxid == key, , drop = FALSE]
      if (nrow(m) > 0L && "species_name" %in% names(m)) { nm <- as.character(m$species_name[1]); break }
    }
    if (nzchar(nm)) sprintf("taxid %s — %s", key, nm) else paste("taxid", key)
  }

  ggplot(lines, aes(x = .data$position, y = .data$pct)) +
    geom_rect(data = spans, inherit.aes = FALSE,
              aes(xmin = .data$xmin, xmax = .data$xmax, ymin = -Inf, ymax = Inf,
                  fill = .data$end),
              alpha = 0.08, show.legend = FALSE) +
    geom_hline(data = plats, aes(yintercept = .data$yint, colour = .data$end),
               linetype = "dotted", linewidth = 0.35, alpha = 0.7,
               show.legend = FALSE) +
    geom_line(aes(colour = .data$end, linetype = .data$sample), linewidth = 0.7) +
    geom_point(aes(colour = .data$end), size = 0.9, show.legend = FALSE) +
    geom_text(data = ann, inherit.aes = FALSE,
              aes(x = -Inf, y = Inf, label = .data$label, colour = .data$end),
              hjust = -0.04, vjust = 1.3, size = 2.3, show.legend = FALSE) +
    # no colour guide: each facet holds one end and the strip already names it
    scale_colour_manual(values = end_pal, guide = "none") +
    scale_fill_manual(values = end_pal, guide = "none") +
    scale_linetype_manual(values = lty, name = NULL,
                          guide = if (n_samp > 1L) "legend" else "none") +
    # Stacked in one column on a SHARED y range, and not anchored at zero.
    # Anchoring at zero left the curve using only ~20% of the panel for a
    # taxon with a high baseline unclassified rate. The range stays shared
    # across ends because 5' vs 3' asymmetry is diagnostic and free scales
    # would make both panels fill their box identically, hiding it.
    facet_grid(end ~ ., switch = "y", labeller = labeller(end = END_LABELS)) +
    coord_cartesian(xlim = c(-0.5, args$max_pos - 0.5)) +
    scale_y_continuous(expand = expansion(mult = 0.08)) +
    labs(
      title = paste("aDNA damage profiles —", page_title),
      x = "Position from read end",
      y = "Unclassified k-mers (%)",
      caption = paste(
        "Rising tail beyond plateau reflects read-length artifact",
        "(short reads: distant positions approach the opposite end)."
      )
    ) +
    theme_screen() +
    theme(legend.position = "top", legend.direction = "vertical",
          panel.spacing.y = unit(0.5, "lines"), strip.placement = "outside")
}

# ---------------------------------------------------------------------------
# Write multi-page PDF
# ---------------------------------------------------------------------------

open_pdf_device(args$output, width = 7.5, height = 6.4)

pages <- 0L
for (k in keys) {
  p <- build_page(k)
  if (is.null(p)) next
  print(p)
  pages <- pages + 1L
}
if (pages == 0L) {
  print(
    ggplot() +
      annotate("text", x = 0.5, y = 0.6, label = "No damage-profile plots were generated",
               size = 4.6, fontface = "bold") +
      annotate("text", x = 0.5, y = 0.42,
               label = "Selected taxa had no matching rows in the profile table.",
               size = 3.4) +
      xlim(0, 1) + ylim(0, 1) + theme_void()
  )
  pages <- 1L
}
invisible(dev.off())
message("Saved ", args$output, " (", pages, " page(s))")
